ssh/controlmaster.go

```diff
@@ -0,0 +1,233 @@
+func DialControlMaster(path string) (*Client, error) {
+	conn, err := net.Dial("unix", path)
+	if err != nil {
+		return nil, err
+	}
+	c, chans, reqs, err := NewControlMasterClientConn(conn)
+	if err != nil {
+		return nil, err
+	}
+	return NewClient(c, chans, reqs), nil
+}
+
+// NewControlMasterClientConn establishes a ControlMaster connection
+// using c as the underlying transport. The Request and NewChannel
+// channels must be serviced or the connection will hang.
+func NewControlMasterClientConn(c net.Conn) (Conn, <-chan NewChannel, <-chan *Request, error) {
+	// We never populate transport, just mux since there is no handshake needed
+	conn := &connection{
+		sshConn: sshConn{conn: c},
+	}
+
+	// Instead of calling clientHandshake, call our own setup function
+	if err := setupControlMaster(&controlMasterConnection{
+		rw: c,
+	}); err != nil {
+		return nil, nil, nil, err
+	}
+
+	// Instead of using transport we use our own impl packetConn instead
+	conn.mux = newMux(&controlMasterProxyConnection{
+		rwc: c,
+	})
+	return conn, conn.mux.incomingChannels, conn.mux.incomingRequests, nil
+}
+
+// controlMasterConnection implements the control master protocol
+type controlMasterConnection struct {
+	rw io.ReadWriter
+}
+
+// writePacket prefixes the length to the packet and writes it
+func (c *controlMasterConnection) writePacket(packet interface{}) error {
+	// Marshal the packet as standard SSH packets
+	msg := Marshal(packet)
+	// We need to calculate the length field
+	msgLenBytes := make([]byte, 4)
+	binary.BigEndian.PutUint32(msgLenBytes, uint32(len(msg)))
+	msg = append(msgLenBytes, msg...)
+	n, err := c.rw.Write(msg)
+	if err != nil {
+		return err
+	}
+	if n != len(msg) {
+		return errors.New("Unable to write all bytes in message")
+	}
+	return nil
+}
+
+// readPacket reads a packet based on an initial length value
+func (c *controlMasterConnection) readPacket(packet interface{}) error {
+	// Read the response length
+	respLenBuf := make([]byte, 4)
+	n1, err := c.rw.Read(respLenBuf)
+	if err != nil {
+		return err
+	}
+	if n1 != 4 {
+		return errors.New("Unable to read entire response packet length")
+	}
+	respLen := binary.BigEndian.Uint32(respLenBuf)
+	// Then read the response
+	resp := make([]byte, respLen)
+	n2, err := c.rw.Read(resp)
+	if err != nil {
+		return err
+	}
+	if uint32(n2) != respLen {
+		return errors.New("Unable to read entire response packet")
+	}
+	if err := Unmarshal(resp, packet); err != nil {
+		return err
+	}
+	return nil
+}
+
+// setupControlMaster gets a CM session setup and switches to proxy mode
+func setupControlMaster(c *controlMasterConnection) error {
+	// Write our own hello with version
+	if err := c.writePacket(&controlMasterHelloMessage{
+		// MUX_MSG_HELLO
+		PacketType:      0x00000001,
+		ProtocolVersion: 4,
+	}); err != nil {
+		return err
+	}
+	// Read the hello message
+	var hello controlMasterHelloMessage
+	if err := c.readPacket(&hello); err != nil {
+		return err
+	}
+	// MUX_MSG_HELLO
```

Comment:
If you make constants for the magic numbers, you don't need these kinds of comments.  You can comment each Go constant with the 'normal' constant identifier in the const block if you want to make it more greppable.